// naive-ui/es/notification/src/NotificationContainer.js
import { h, defineComponent, inject } from 'vue';
import { NScrollbar } from '../../_internal';
import { notificationProviderInjectionKey } from './context';
export const NotificationContainer = defineComponent({
    name: 'NotificationContainer',
    props: {
        scrollable: {
            type: Boolean,
            required: true
        },
        placement: {
            type: String,
            required: true
        }
    },
    setup() {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const { mergedThemeRef, mergedClsPrefixRef } = inject(notificationProviderInjectionKey);
        return {
            mergedTheme: mergedThemeRef,
            mergedClsPrefix: mergedClsPrefixRef
        };
    },
    render() {
        const { $slots, scrollable, mergedClsPrefix, mergedTheme, placement } = this;
        return (h("div", { class: [
                `${mergedClsPrefix}-notification-container`,
                scrollable && `${mergedClsPrefix}-notification-container--scrollable`,
                `${mergedClsPrefix}-notification-container--${placement}`
            ] }, scrollable ? (h(NScrollbar, { theme: mergedTheme.peers.Scrollbar, themeOverrides: mergedTheme.peerOverrides.Scrollbar }, $slots)) : ($slots)));
    }
});
