// vueuc/lib/resize-observer/src/VResizeObserver.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const delegate_1 = __importDefault(require("./delegate"));
const shared_1 = require("../../shared");
exports.default = (0, vue_1.defineComponent)({
    name: 'ResizeObserver',
    props: {
        onResize: Function
    },
    setup(props) {
        return {
            registered: false,
            handleResize(entry) {
                const { onResize } = props;
                if (onResize !== undefined)
                    onResize(entry);
            }
        };
    },
    mounted() {
        const el = this.$el;
        if (el === undefined) {
            (0, shared_1.warn)('resize-observer', '$el does not exist.');
            return;
        }
        if (el.nextElementSibling !== el.nextSibling) {
            if (el.nodeType === 3 && el.nodeValue !== '') {
                (0, shared_1.warn)('resize-observer', '$el can not be observed (it may be a text node).');
                return;
            }
        }
        if (el.nextElementSibling !== null) {
            delegate_1.default.registerHandler(el.nextElementSibling, this.handleResize);
            this.registered = true;
        }
    },
    beforeUnmount() {
        if (this.registered) {
            delegate_1.default.unregisterHandler(this.$el.nextElementSibling);
        }
    },
    render() {
        return (0, vue_1.renderSlot)(this.$slots, 'default');
    }
});
