// naive-ui/lib/form/src/styles/form-item.cssr.js
"use strict";

var __importDefault = this && this.__importDefault || function (mod) {
  return mod && mod.__esModule ? mod : {
    "default": mod
  };
};

Object.defineProperty(exports, "__esModule", {
  value: true
});

const cssr_1 = require("../../../_utils/cssr");

const fade_down_cssr_1 = __importDefault(require("../../../_styles/transitions/fade-down.cssr")); // vars:
// --n-line-height
// --n-blank-height
// --n-feedback-padding
// --n-feedback-font-size
// --n-label-font-size-left
// --n-label-font-size-top
// --n-label-height
// --n-label-padding
// --n-asterisk-color
// --n-label-text-color
// --n-bezier
// --n-feedback-text-color
// --n-feedback-text-color-warning
// --n-feedback-text-color-error
// --n-label-text-align
// --n-label-padding


exports.default = (0, cssr_1.cB)('form-item', {
  display: 'grid',
  lineHeight: 'var(--n-line-height)'
}, [(0, cssr_1.cB)('form-item-label', `
 grid-area: label;
 align-items: center;
 line-height: 1.25;
 text-align: var(--n-label-text-align);
 font-size: var(--n-label-font-size);
 height: var(--n-label-height);
 padding: var(--n-label-padding);
 color: var(--n-label-text-color);
 transition: color .3s var(--n-bezier);
 box-sizing: border-box;
 `, [(0, cssr_1.cE)('asterisk', `
 color: var(--n-asterisk-color);
 transition: color .3s var(--n-bezier);
 `), (0, cssr_1.cE)('asterisk-placeholder', `
 visibility: hidden; 
 `)]), (0, cssr_1.cB)('form-item-blank', {
  gridArea: 'blank',
  minHeight: 'var(--n-blank-height)'
}), (0, cssr_1.cM)('left-labelled', `
 grid-template-areas:
 "label blank"
 "label feedback";
 grid-template-columns: auto minmax(0, 1fr);
 `, [(0, cssr_1.cB)('form-item-label', `
 height: var(--n-blank-height);
 line-height: var(--n-blank-height);
 box-sizing: border-box;
 white-space: nowrap;
 flex-shrink: 0;
 flex-grow: 0;
 `)]), (0, cssr_1.cM)('top-labelled', `
 grid-template-areas:
 "label"
 "blank"
 "feedback";
 grid-template-rows: var(--n-label-height) 1fr;
 grid-template-columns: minmax(0, 100%);
 `, [(0, cssr_1.cM)('no-label', `
 grid-template-areas:
 "blank"
 "feedback";
 grid-template-rows: 1fr;
 `), (0, cssr_1.cB)('form-item-label', {
  display: 'flex',
  alignItems: 'flex-end',
  justifyContent: 'var(--n-label-text-align)'
})]), (0, cssr_1.cB)('form-item-blank', `
 box-sizing: border-box;
 display: flex;
 align-items: center;
 position: relative;
 `), (0, cssr_1.cB)('form-item-feedback-wrapper', `
 grid-area: feedback;
 box-sizing: border-box;
 min-height: var(--n-feedback-height);
 font-size: var(--n-feedback-font-size);
 line-height: 1.25;
 transform-origin: top left;
 `, [(0, cssr_1.c)('&:not(:empty)', `
 padding: var(--n-feedback-padding);
 `), (0, cssr_1.cB)('form-item-feedback', {
  transition: 'color .3s var(--n-bezier)',
  color: 'var(--n-feedback-text-color)'
}, [(0, cssr_1.cM)('warning', {
  color: 'var(--n-feedback-text-color-warning)'
}), (0, cssr_1.cM)('error', {
  color: 'var(--n-feedback-text-color-error)'
}), (0, fade_down_cssr_1.default)({
  fromOffset: '-3px',
  enterDuration: '.3s',
  leaveDuration: '.2s'
})])])]);
