// naive-ui/es/typography/src/text.js
import { h, defineComponent, computed } from 'vue';
import { useCompitable } from 'vooks';
import { useConfig, useTheme } from '../../_mixins';
import { warn, createKey } from '../../_utils';
import { typographyLight } from '../styles';
import style from './styles/text.cssr';
const textProps = Object.assign(Object.assign({}, useTheme.props), { code: Boolean, type: {
        type: String,
        default: 'default'
    }, delete: Boolean, strong: Boolean, italic: Boolean, underline: Boolean, depth: [String, Number], tag: String, 
    // deprecated
    as: {
        type: String,
        validator: () => {
            if (process.env.NODE_ENV !== 'production') {
                warn('text', '`as` is deprecated, please use `tag` instead.');
            }
            return true;
        },
        default: undefined
    } });
export default defineComponent({
    name: 'Text',
    props: textProps,
    setup(props) {
        const { mergedClsPrefixRef } = useConfig(props);
        const themeRef = useTheme('Typography', '-text', style, typographyLight, props, mergedClsPrefixRef);
        return {
            mergedClsPrefix: mergedClsPrefixRef,
            compitableTag: useCompitable(props, ['as', 'tag']),
            cssVars: computed(() => {
                const { depth, type } = props;
                const textColorKey = type === 'default'
                    ? depth === undefined
                        ? 'textColor'
                        : `textColor${depth}Depth`
                    : createKey('textColor', type);
                const { common: { fontWeightStrong, fontFamilyMono, cubicBezierEaseInOut }, self: { codeTextColor, codeBorderRadius, codeColor, codeBorder, [textColorKey]: textColor } } = themeRef.value;
                return {
                    '--n-bezier': cubicBezierEaseInOut,
                    '--n-text-color': textColor,
                    '--n-font-weight-strong': fontWeightStrong,
                    '--n-font-famliy-mono': fontFamilyMono,
                    '--n-code-border-radius': codeBorderRadius,
                    '--n-code-text-color': codeTextColor,
                    '--n-code-color': codeColor,
                    '--n-code-border': codeBorder
                };
            })
        };
    },
    render() {
        var _a, _b;
        const { mergedClsPrefix } = this;
        const textClass = [
            `${mergedClsPrefix}-text`,
            {
                [`${mergedClsPrefix}-text--code`]: this.code,
                [`${mergedClsPrefix}-text--delete`]: this.delete,
                [`${mergedClsPrefix}-text--strong`]: this.strong,
                [`${mergedClsPrefix}-text--italic`]: this.italic,
                [`${mergedClsPrefix}-text--underline`]: this.underline
            }
        ];
        const children = (_b = (_a = this.$slots).default) === null || _b === void 0 ? void 0 : _b.call(_a);
        return this.code ? (h("code", { class: textClass, style: this.cssVars }, this.delete ? h("del", null, children) : children)) : this.delete ? (h("del", { class: textClass, style: this.cssVars }, children)) : (h(this.compitableTag || 'span', { class: textClass, style: this.cssVars }, children));
    }
});
