// naive-ui/lib/statistic/src/Statistic.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const _mixins_1 = require("../../_mixins");
const styles_1 = require("../styles");
const index_cssr_1 = __importDefault(require("./styles/index.cssr"));
const statisticProps = Object.assign(Object.assign({}, _mixins_1.useTheme.props), { tabularNums: Boolean, label: String, value: [String, Number] });
exports.default = (0, vue_1.defineComponent)({
    name: 'Statistic',
    props: statisticProps,
    setup(props) {
        const { mergedClsPrefixRef } = (0, _mixins_1.useConfig)(props);
        const themeRef = (0, _mixins_1.useTheme)('Statistic', '-statistic', index_cssr_1.default, styles_1.statisticLight, props, mergedClsPrefixRef);
        return {
            mergedClsPrefix: mergedClsPrefixRef,
            cssVars: (0, vue_1.computed)(() => {
                const { self: { labelFontWeight, valueFontWeight, valuePrefixTextColor, labelTextColor, valueSuffixTextColor, valueTextColor, labelFontSize }, common: { cubicBezierEaseInOut } } = themeRef.value;
                return {
                    '--n-bezier': cubicBezierEaseInOut,
                    '--n-label-font-size': labelFontSize,
                    '--n-label-font-weight': labelFontWeight,
                    '--n-label-text-color': labelTextColor,
                    '--n-value-font-weight': valueFontWeight,
                    '--n-value-prefix-text-color': valuePrefixTextColor,
                    '--n-value-suffix-text-color': valueSuffixTextColor,
                    '--n-value-text-color': valueTextColor
                };
            })
        };
    },
    render() {
        var _a;
        const { $slots, mergedClsPrefix } = this;
        return ((0, vue_1.h)("div", { class: `${mergedClsPrefix}-statistic`, style: this.cssVars },
            (0, vue_1.h)("div", { class: `${mergedClsPrefix}-statistic__label` }, this.label || ((_a = $slots.label) === null || _a === void 0 ? void 0 : _a.call($slots))),
            (0, vue_1.h)("div", { class: `${mergedClsPrefix}-statistic-value`, style: {
                    fontVariantNumeric: this.tabularNums ? 'tabular-nums' : ''
                } },
                $slots.prefix ? ((0, vue_1.h)("span", { class: `${mergedClsPrefix}-statistic-value__prefix` }, $slots.prefix())) : null,
                this.value !== undefined ? ((0, vue_1.h)("span", { class: `${mergedClsPrefix}-statistic-value__content` }, this.value)) : ((0, vue_1.h)("span", { class: `${mergedClsPrefix}-statistic-value__content` }, $slots)),
                $slots.suffix ? ((0, vue_1.h)("span", { class: `${mergedClsPrefix}-statistic-value__suffix` }, $slots.suffix())) : null)));
    }
});
