// naive-ui/es/badge/src/Badge.js
import { h, computed, onMounted, ref, defineComponent, Transition } from 'vue';
import { useConfig, useTheme } from '../../_mixins';
import { NBaseSlotMachine, NBaseWave } from '../../_internal';
import { createKey, getTitleAttribute } from '../../_utils';
import { badgeLight } from '../styles';
import style from './styles/index.cssr';
const badgeProps = Object.assign(Object.assign({}, useTheme.props), { value: [String, Number], max: Number, dot: {
        type: Boolean,
        default: false
    }, type: {
        type: String,
        default: 'default'
    }, show: {
        type: Boolean,
        default: true
    }, showZero: {
        type: Boolean,
        default: false
    }, processing: {
        type: Boolean,
        default: false
    }, color: String });
export default defineComponent({
    name: 'Badge',
    props: badgeProps,
    setup(props) {
        const { mergedClsPrefixRef } = useConfig(props);
        const themeRef = useTheme('Badge', '-badge', style, badgeLight, props, mergedClsPrefixRef);
        const appearedRef = ref(false);
        const handleAfterEnter = () => {
            appearedRef.value = true;
        };
        const handleAfterLeave = () => {
            appearedRef.value = false;
        };
        const showBadgeRef = computed(() => {
            return (props.show &&
                (props.dot ||
                    (props.value !== undefined && !(!props.showZero && props.value <= 0))));
        });
        onMounted(() => {
            if (showBadgeRef.value)
                appearedRef.value = true;
        });
        return {
            mergedClsPrefix: mergedClsPrefixRef,
            appeared: ref(false),
            showBadge: showBadgeRef,
            handleAfterEnter,
            handleAfterLeave,
            cssVars: computed(() => {
                const { type, color: propColor } = props;
                const { common: { cubicBezierEaseInOut, cubicBezierEaseOut }, self: { [createKey('color', type)]: color, fontFamily, fontSize } } = themeRef.value;
                return {
                    '--n-font-size': fontSize,
                    '--n-font-family': fontFamily,
                    '--n-color': propColor || color,
                    '--n-ripple-color': propColor || color,
                    '--n-bezier': cubicBezierEaseInOut,
                    '--n-ripple-bezier': cubicBezierEaseOut
                };
            })
        };
    },
    render() {
        var _a;
        const { mergedClsPrefix, $slots } = this;
        return (h("div", { class: [
                `${mergedClsPrefix}-badge`,
                {
                    [`${mergedClsPrefix}-badge--dot`]: this.dot,
                    [`${mergedClsPrefix}-badge--as-is`]: !$slots.default
                }
            ], style: this.cssVars }, (_a = $slots.default) === null || _a === void 0 ? void 0 :
            _a.call($slots),
            h(Transition, { name: "fade-in-scale-up-transition", onAfterEnter: this.handleAfterEnter, onAfterLeave: this.handleAfterLeave }, {
                default: () => this.showBadge ? (h("sup", { class: `${mergedClsPrefix}-badge-sup`, title: getTitleAttribute(this.value) },
                    !this.dot ? (h(NBaseSlotMachine, { clsPrefix: mergedClsPrefix, appeared: this.appeared, max: this.max, value: this.value })) : null,
                    this.processing ? (h(NBaseWave, { clsPrefix: mergedClsPrefix })) : null)) : null
            })));
    }
});
