// naive-ui/lib/timeline/src/TimelineItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const _utils_1 = require("../../_utils");
const Timeline_1 = require("./Timeline");
const timelineItemProps = {
    time: [String, Number],
    title: String,
    content: String,
    color: String,
    type: {
        type: String,
        default: 'default'
    }
};
exports.default = (0, vue_1.defineComponent)({
    name: 'TimelineItem',
    props: timelineItemProps,
    setup(props) {
        const NTimeline = (0, vue_1.inject)(Timeline_1.timelineInjectionKey);
        if (!NTimeline) {
            (0, _utils_1.throwError)('timeline-item', '`n-timeline-item` must be placed inside `n-timeline`.');
        }
        return {
            mergedClsPrefix: NTimeline.mergedClsPrefixRef,
            cssVars: (0, vue_1.computed)(() => {
                const { props: { size, iconSize: iconSizeProp }, mergedThemeRef } = NTimeline;
                const { type } = props;
                const { self: { titleTextColor, contentTextColor, metaTextColor, lineColor, titleFontWeight, contentFontSize, [(0, _utils_1.createKey)('iconSize', size)]: iconSize, [(0, _utils_1.createKey)('titleMargin', size)]: titleMargin, [(0, _utils_1.createKey)('titleFontSize', size)]: titleFontSize, [(0, _utils_1.createKey)('circleBorder', type)]: circleBorder, [(0, _utils_1.createKey)('iconColor', type)]: iconColor }, common: { cubicBezierEaseInOut } } = mergedThemeRef.value;
                return {
                    '--n-bezier': cubicBezierEaseInOut,
                    '--n-circle-border': circleBorder,
                    '--n-icon-color': iconColor,
                    '--n-content-font-size': contentFontSize,
                    '--n-content-text-color': contentTextColor,
                    '--n-line-color': lineColor,
                    '--n-meta-text-color': metaTextColor,
                    '--n-title-font-size': titleFontSize,
                    '--n-title-font-weight': titleFontWeight,
                    '--n-title-margin': titleMargin,
                    '--n-title-text-color': titleTextColor,
                    '--n-icon-size': (0, _utils_1.formatLength)(iconSizeProp) || iconSize
                };
            })
        };
    },
    render() {
        const { mergedClsPrefix, color, $slots } = this;
        return ((0, vue_1.h)("div", { class: [
                `${mergedClsPrefix}-timeline-item`,
                `${mergedClsPrefix}-timeline-item--${this.type}-type`
            ], style: this.cssVars },
            (0, vue_1.h)("div", { class: `${mergedClsPrefix}-timeline-item-timeline` },
                (0, vue_1.h)("div", { class: `${mergedClsPrefix}-timeline-item-timeline__line` }),
                $slots.icon ? ((0, vue_1.h)("div", { class: `${mergedClsPrefix}-timeline-item-timeline__icon`, style: { color: color } }, $slots.icon())) : ((0, vue_1.h)("div", { class: `${mergedClsPrefix}-timeline-item-timeline__circle`, style: { borderColor: color } }))),
            (0, vue_1.h)("div", { class: `${mergedClsPrefix}-timeline-item-content` },
                this.title || $slots.header ? ((0, vue_1.h)("div", { class: `${mergedClsPrefix}-timeline-item-content__title` }, $slots.header ? $slots.header() : this.title)) : null,
                (0, vue_1.h)("div", { class: `${mergedClsPrefix}-timeline-item-content__content` }, $slots.default ? $slots.default() : this.content),
                (0, vue_1.h)("div", { class: `${mergedClsPrefix}-timeline-item-content__meta` }, $slots.footer ? $slots.footer() : this.time))));
    }
});
