// naive-ui/es/button/src/ButtonGroup.js
import { h, defineComponent, provide } from 'vue';
import { useConfig, useStyle } from '../../_mixins';
import { createInjectionKey } from '../../_utils';
import style from './styles/button-group.cssr';
export const buttonGroupInjectionKey = createInjectionKey('n-button-group');
const buttonGroupProps = {
    size: {
        type: String,
        default: undefined
    },
    vertical: Boolean
};
export default defineComponent({
    name: 'ButtonGroup',
    props: buttonGroupProps,
    setup(props) {
        const { mergedClsPrefixRef } = useConfig(props);
        useStyle('-button-group', style, mergedClsPrefixRef);
        provide(buttonGroupInjectionKey, props);
        return {
            mergedClsPrefix: mergedClsPrefixRef
        };
    },
    render() {
        const { mergedClsPrefix } = this;
        return (h("div", { class: [
                `${mergedClsPrefix}-button-group`,
                this.vertical && `${mergedClsPrefix}-button-group--vertical`
            ], role: "group" }, this.$slots));
    }
});
