// naive-ui/lib/typography/src/a.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const _mixins_1 = require("../../_mixins");
const styles_1 = require("../styles");
const a_cssr_1 = __importDefault(require("./styles/a.cssr"));
const aProps = Object.assign({}, _mixins_1.useTheme.props);
exports.default = (0, vue_1.defineComponent)({
    name: 'A',
    props: aProps,
    setup(props) {
        const { mergedClsPrefixRef } = (0, _mixins_1.useConfig)(props);
        const themeRef = (0, _mixins_1.useTheme)('Typography', '-a', a_cssr_1.default, styles_1.typographyLight, props, mergedClsPrefixRef);
        return {
            mergedClsPrefix: mergedClsPrefixRef,
            cssVars: (0, vue_1.computed)(() => {
                const { common: { cubicBezierEaseInOut }, self: { aTextColor } } = themeRef.value;
                return {
                    '--n-text-color': aTextColor,
                    '--n-bezier': cubicBezierEaseInOut
                };
            })
        };
    },
    render() {
        return ((0, vue_1.h)("a", { class: `${this.mergedClsPrefix}-a`, style: this.cssVars }, this.$slots));
    }
});
