// naive-ui/es/statistic/src/Statistic.js
import { defineComponent, computed, h } from 'vue';
import { useConfig, useTheme } from '../../_mixins';
import { statisticLight } from '../styles';
import style from './styles/index.cssr';
const statisticProps = Object.assign(Object.assign({}, useTheme.props), { tabularNums: Boolean, label: String, value: [String, Number] });
export default defineComponent({
    name: 'Statistic',
    props: statisticProps,
    setup(props) {
        const { mergedClsPrefixRef } = useConfig(props);
        const themeRef = useTheme('Statistic', '-statistic', style, statisticLight, props, mergedClsPrefixRef);
        return {
            mergedClsPrefix: mergedClsPrefixRef,
            cssVars: computed(() => {
                const { self: { labelFontWeight, valueFontWeight, valuePrefixTextColor, labelTextColor, valueSuffixTextColor, valueTextColor, labelFontSize }, common: { cubicBezierEaseInOut } } = themeRef.value;
                return {
                    '--n-bezier': cubicBezierEaseInOut,
                    '--n-label-font-size': labelFontSize,
                    '--n-label-font-weight': labelFontWeight,
                    '--n-label-text-color': labelTextColor,
                    '--n-value-font-weight': valueFontWeight,
                    '--n-value-prefix-text-color': valuePrefixTextColor,
                    '--n-value-suffix-text-color': valueSuffixTextColor,
                    '--n-value-text-color': valueTextColor
                };
            })
        };
    },
    render() {
        var _a;
        const { $slots, mergedClsPrefix } = this;
        return (h("div", { class: `${mergedClsPrefix}-statistic`, style: this.cssVars },
            h("div", { class: `${mergedClsPrefix}-statistic__label` }, this.label || ((_a = $slots.label) === null || _a === void 0 ? void 0 : _a.call($slots))),
            h("div", { class: `${mergedClsPrefix}-statistic-value`, style: {
                    fontVariantNumeric: this.tabularNums ? 'tabular-nums' : ''
                } },
                $slots.prefix ? (h("span", { class: `${mergedClsPrefix}-statistic-value__prefix` }, $slots.prefix())) : null,
                this.value !== undefined ? (h("span", { class: `${mergedClsPrefix}-statistic-value__content` }, this.value)) : (h("span", { class: `${mergedClsPrefix}-statistic-value__content` }, $slots)),
                $slots.suffix ? (h("span", { class: `${mergedClsPrefix}-statistic-value__suffix` }, $slots.suffix())) : null)));
    }
});
