// naive-ui/lib/typography/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NText = exports.NLi = exports.NOl = exports.NUl = exports.NHr = exports.NBlockquote = exports.NP = exports.NA = exports.NH6 = exports.NH5 = exports.NH4 = exports.NH3 = exports.NH2 = exports.NH1 = void 0;
var headers_1 = require("./src/headers");
Object.defineProperty(exports, "NH1", { enumerable: true, get: function () { return headers_1.NH1; } });
Object.defineProperty(exports, "NH2", { enumerable: true, get: function () { return headers_1.NH2; } });
Object.defineProperty(exports, "NH3", { enumerable: true, get: function () { return headers_1.NH3; } });
Object.defineProperty(exports, "NH4", { enumerable: true, get: function () { return headers_1.NH4; } });
Object.defineProperty(exports, "NH5", { enumerable: true, get: function () { return headers_1.NH5; } });
Object.defineProperty(exports, "NH6", { enumerable: true, get: function () { return headers_1.NH6; } });
var a_1 = require("./src/a");
Object.defineProperty(exports, "NA", { enumerable: true, get: function () { return __importDefault(a_1).default; } });
var p_1 = require("./src/p");
Object.defineProperty(exports, "NP", { enumerable: true, get: function () { return __importDefault(p_1).default; } });
var blockquote_1 = require("./src/blockquote");
Object.defineProperty(exports, "NBlockquote", { enumerable: true, get: function () { return __importDefault(blockquote_1).default; } });
var hr_1 = require("./src/hr");
Object.defineProperty(exports, "NHr", { enumerable: true, get: function () { return __importDefault(hr_1).default; } });
var ul_1 = require("./src/ul");
Object.defineProperty(exports, "NUl", { enumerable: true, get: function () { return __importDefault(ul_1).default; } });
var ol_1 = require("./src/ol");
Object.defineProperty(exports, "NOl", { enumerable: true, get: function () { return __importDefault(ol_1).default; } });
var li_1 = require("./src/li");
Object.defineProperty(exports, "NLi", { enumerable: true, get: function () { return __importDefault(li_1).default; } });
var text_1 = require("./src/text");
Object.defineProperty(exports, "NText", { enumerable: true, get: function () { return __importDefault(text_1).default; } });
